### ISSUE TITLE
Pretty-printing a (nested) instance of a class with many static constants of its own type produces an explosion of repeated output, rendering the pretty-print unusable

### ISSUE THREAD

**Issue by andersthorbeck** [snowe2010/pretty-print#23]
Version: `com.tylerthrailkill.helpers:pretty-print:2.0.2`

I've run into an issue pretty-printing an object that deeply nested contains an instance of `org.springframework.http.MediaType`.

What happens is that `pretty-print` prints every single `static final` constant within `MediaType`, of which there are exactly 50. However, 23 of these constants are themselves instances of `MediaType`, which are again recursively pretty-printed.

To its credit, `pretty-print` does recognize cyclic references and halt infinite recursion. However, this only happens when the reference is literally cyclic, i.e. is already further _up_ the call (print) stack. It does _not_ prevent the same field from being printed again elsewhere in the print "tree", which for 23 constants produces an absolutely enourmous print tree for just a single `MediaType` instance.

The start of the pretty-print looks as follows. It was still running when I cancelled it after **~700k lines**, all spawned from the single `MediaType` instance.
```
                 ...
                 description = MarkupProperty(
                   contentType = MediaType(
                     serialVersionUID = 2069937152339670231
                     ALL = MediaType(
                       serialVersionUID = 2069937152339670231
                       ALL = cyclic reference detected for 1100568629
                       ALL_VALUE = "*/*"
                       APPLICATION_ATOM_XML = MediaType(
                         serialVersionUID = 2069937152339670231
                         ALL = cyclic reference detected for 1100568629
                         ALL_VALUE = "*/*"
                         APPLICATION_ATOM_XML = cyclic reference detected for 691278472
                         APPLICATION_ATOM_XML_VALUE = "application/atom+xml"
                         APPLICATION_FORM_URLENCODED = MediaType(
                           serialVersionUID = 2069937152339670231
                           ALL = cyclic reference detected for 1100568629
                           ALL_VALUE = "*/*"
                           APPLICATION_ATOM_XML = cyclic reference detected for 691278472
                           APPLICATION_ATOM_XML_VALUE = "application/atom+xml"
                           APPLICATION_FORM_URLENCODED = cyclic reference detected for 1509426585
                           APPLICATION_FORM_URLENCODED_VALUE = "application/x-www-form-urlencoded"
                           APPLICATION_JSON = MediaType(
                             serialVersionUID = 2069937152339670231
                             ALL = cyclic reference detected for 1100568629
                             ALL_VALUE = "*/*"
                             APPLICATION_ATOM_XML = cyclic reference detected for 691278472
                             APPLICATION_ATOM_XML_VALUE = "application/atom+xml"
                             APPLICATION_FORM_URLENCODED = cyclic reference detected for 1509426585
                             APPLICATION_FORM_URLENCODED_VALUE = "application/x-www-form-urlencoded"
                             APPLICATION_JSON = cyclic reference detected for 339219747
                             APPLICATION_JSON_VALUE = "application/json"
                             APPLICATION_JSON_UTF8 = MediaType(
                               serialVersionUID = 2069937152339670231
                               ALL = cyclic reference detected for 1100568629
                               ALL_VALUE = "*/*"
                               ...
                               ...
                               ...
```

To illustrate the issue, imagine pretty printing the following simplified example.
```java
public class FooBar {
  public static final FooBar FOO;
  public static final FooBar BAR;
  public static final FooBar BAZ;
}
```

The pretty print of `foobar: FooBar` with the current behaviour would be
```
foobar = FooBar(
  FOO = FooBar(
    FOO = cyclic
    BAR = FooBar(
      FOO = cyclic
      BAR = cyclic
      BAZ = FooBar(
        FOO = cyclic
        BAR = cyclic
        BAZ = cyclic
      )
    )
    BAZ = FooBar(
      FOO = cyclic
      BAR = FooBar(
        FOO = cyclic
        BAR = cyclic
        BAZ = cyclic
      )
      BAZ = cyclic
    )
  )
  BAR = FooBar(
    FOO = FooBar(
      FOO = cyclic
      BAR = cyclic
      BAZ = FooBar(
        FOO = cyclic
        BAR = cyclic
        BAZ = cyclic
      )
    )
    BAR = cyclic
    BAZ = FooBar(
      FOO = FooBar(
        FOO = cyclic
        BAR = cyclic
        BAZ = cyclic
      )
      BAR = cyclic
      BAZ = cyclic
    )
  )
  BAZ = FooBar(
    FOO = FooBar(
      FOO = cyclic
      BAR = FooBar(
        FOO = cyclic
        BAR = cyclic
        BAZ = cyclic
      )
      BAZ = cyclic
    )
    BAR = FooBar(
      FOO = FooBar(
        FOO = cyclic
        BAR = cyclic
        BAZ = cyclic
      )
      BAR = cyclic
      BAZ = cyclic
    )
    BAZ = cyclic
  )
)
```

That's 65 lines pretty printed, from a single object whose class contains 3 static constants of its own type.
With `MediaType`, which has 23 such constants instead of 3, the output is exponentially bigger.

**TL;DR:** Pretty-printing objects with many static constants of its own type is unfeasible. 

Of course, this is an example based on a Java-class, as that is where the issue arose for me. I have not tested whether similar issues arise with kotlin. However, since Kotlin still very often will need to use Java classes from external frameworks and libraries, I still reckon this is an important issue to tackle.

It seems to me that one natural solution is to just not print static constants. Why would we be interested in printing out, for every single instance of a type, all the fields which are not instance-specific?

**Comment by snowe2010** [comment]
Hmm, yes this is a pretty big issue, since this occurs with such a basic example, and especially because this occurs with such a common object such as `MediaType`.

A simple solution is to not print statics, maybe with an option to print statics if you want. I'm looking into this, not sure how long it will take me to have a fix. I've already got a simple fix for just removing statics from being printed, but would like a fleshed out solution.

**Comment by snowe2010** [comment]
Another option here might be detecting cycles quicker. Cycle detection seems to be a function of the number of fields, e.g. if there are 3 fields, the full cycle isn't detected until we get to depth 3, 4 fields = depth 4, etc.

**Comment by ndon55555** [comment]
I do see Anders' point that maybe mixing in static values alongside instance values could be a bit confusing. Still though, it might be nice to know what values static types hold. Here's one idea:

If an object P of type T being printed has a static, non-primitive field C of the same type, then when pretty printing C, do not print its static fields of type T.

Adjusting @andersthorbeck's example slightly to include an instance variable, suppose we have

```java
public class FooBar {
  public static final FooBar FOO;
  public static final FooBar BAR;
  public static final FooBar BAZ;
  public int n;
}
```

I would expect a pretty printed output of a `FooBar` instance to look something like:

```
FooBar(
  FOO = FooBar(
    n = 1
  )
  BAR = FooBar(
    n = 3
  )
  BAZ = FooBar(
    n = 2
  )
  n = 7
)
```

We could also mention when we're printing out static variables and when we're going to omit them to clearly distinguish between static and instance variables.

**Comment by snowe2010** [comment]
After looking at this for a good long while, I think it's not going to be feasible (for me) to make `pretty-print` print static members that have the same type as their parent, even if I exclude the static fields of type T per @ndon55555's suggestion. Imagine two Java objects:

```java
public class StaticFieldsObject {
    public static final StaticFieldsObject FOO;
    public StaticFieldsObject2 n;
    
    static {
        FOO = new StaticFieldsObject(StaticFieldsObject2.BAR);
    }

    public StaticFieldsObject(StaticFieldsObject2 i) {
        n = i;
    }
}
```

and 

```java
public class StaticFieldsObject2 {
    public static final StaticFieldsObject2 BAR;
    public StaticFieldsObject n;
    
    static {
        BAR = new StaticFieldsObject2(StaticFieldsObject.FOO);
    }

    public StaticFieldsObject2(StaticFieldsObject i) {
        n = i;
    }
}
```

In this case, what do I print? Currently I can get this output:

```
StaticFieldsObject(
  FOO = static member matches parent. Printing only non-static non-cyclic members (
        n = StaticFieldsObject2(
          BAR = static member matches parent. Printing only non-static non-cyclic members (
                n = null
          )
          n = null
        )[$id=20554286]
        
  n = StaticFieldsObject2(
    BAR = static member matches parent. Printing only non-static non-cyclic members (
          n = null
    )
    n = null
  )[$id=20554286]
)[$id=279497676]
```

but that's pretty much unreadable. I'm going to instead go for something like this output:

```
StaticFieldsObject(
  FOO = StaticFieldsObject.<static cyclic class reference>,
  n = StaticFieldsObject2(
    BAR = StaticFieldsObject.<static cyclic class reference>,
    n = null
  )[$id=800236875]
)[$id=1307997067]
```

What do you think of that? @andersthorbeck and @ndon55555?

**Comment by snowe2010** [comment]
Hm. it doesn't work exactly as above, just due to each static member having a different global id, but this is what I've got so far:

```
StaticFieldsObject(
  FOO = StaticFieldsObject.<static cyclic class reference for 1584361736>
  BAR = StaticFieldsObject.<static cyclic class reference for 1624067520>
  BAZ = StaticFieldsObject.<static cyclic class reference for 2112644667>
  BIZ = StaticFieldsObject.<static cyclic class reference for 517302999>
  BOOZE = StaticFieldsObject.<static cyclic class reference for 828679685>
  n = 0
)
```

I think I'll remove the id and just have it like so:

```
StaticFieldsObject(
  FOO = StaticFieldsObject.<static cyclic class reference>
  BAR = StaticFieldsObject.<static cyclic class reference>
  BAZ = StaticFieldsObject.<static cyclic class reference>
  BIZ = StaticFieldsObject.<static cyclic class reference>
  BOOZE = StaticFieldsObject.<static cyclic class reference>
  n = 0
)
```

Thoughts?

**Comment by snowe2010** [comment]
Released in https://github.com/snowe2010/pretty-print/releases/tag/v2.0.8

You can find the package in [Github Packages](https://github.com/snowe2010/pretty-print/packages/424577?version=v2.0.8), [Bintray](https://bintray.com/snowe/maven/Pretty-Print/v2.0.8#), or [Maven Central](https://search.maven.org/artifact/com.tylerthrailkill.helpers/pretty-print) (might take a while for Maven Central)
